crates/bitwarden-crypto/src/signing/verifying_key.rs

```diff
@@ -0,0 +1,153 @@
+//! A verifying key is the public part of a signature key pair. It is used to verify signatures.
+//!
+//! This implements the lowest layer of the signature module, verifying signatures on raw byte
+//! arrays.
+
+use ciborium::{value::Integer, Value};
+use coset::{
+    iana::{Algorithm, EllipticCurve, EnumI64, KeyOperation, KeyType, OkpKeyParameter},
+    CborSerializable, RegisteredLabel, RegisteredLabelWithPrivate,
+};
+
+use super::{ed25519_verifying_key, key_id, SignatureAlgorithm};
+use crate::{cose::CoseSerializable, error::SignatureError, keys::KeyId, CryptoError};
+
+/// A `VerifyingKey` without the key id. This enum contains a variant for each supported signature
+/// scheme.
+pub(super) enum RawVerifyingKey {
+    Ed25519(ed25519_dalek::VerifyingKey),
+}
+
+/// A verifying key is a public key used for verifying signatures. It can be published to other
+/// users, who can use it to verify that messages were signed by the holder of the corresponding
+/// `SigningKey`.
+pub struct VerifyingKey {
+    pub(super) id: KeyId,
+    pub(super) inner: RawVerifyingKey,
+}
+
+impl VerifyingKey {
+    /// Returns the signature scheme used by the verifying key.
+    pub fn algorithm(&self) -> SignatureAlgorithm {
+        match &self.inner {
+            RawVerifyingKey::Ed25519(_) => SignatureAlgorithm::Ed25519,
+        }
+    }
+
+    /// Verifies the signature of the given data, for the given namespace.
+    /// This should never be used directly, but only through the `verify` method, to enforce
+    /// strong domain separation of the signatures.
+    pub(super) fn verify_raw(&self, signature: &[u8], data: &[u8]) -> Result<(), CryptoError> {
+        match &self.inner {
+            RawVerifyingKey::Ed25519(key) => {
+                let sig = ed25519_dalek::Signature::from_bytes(
+                    signature
+                        .try_into()
+                        .map_err(|_| SignatureError::InvalidSignature)?,
+                );
+                key.verify_strict(data, &sig)
+                    .map_err(|_| SignatureError::InvalidSignature.into())
+            }
+        }
+    }
+}
+
+impl CoseSerializable for VerifyingKey {
+    fn to_cose(&self) -> Result<Vec<u8>, CryptoError> {
+        match &self.inner {
+            RawVerifyingKey::Ed25519(key) => coset::CoseKeyBuilder::new_okp_key()
+                .key_id((&self.id).into())
+                .algorithm(Algorithm::EdDSA)
+                .param(
+                    OkpKeyParameter::Crv.to_i64(),
```

Comment:
We documented this in the other location.

```suggestion
                    OkpKeyParameter::Crv.to_i64(), // Elliptic curve identifier
```